python/pyarrow/tests/test_io.py

```diff
@@ -669,6 +669,57 @@ def test_allocate_buffer_resizable():
     assert buf.size == 200
 
 
+def test_non_cpu_buffer(pickle_module):
+    cuda = pytest.importorskip("pyarrow.cuda")
+    ctx = cuda.Context(0)
+
+    arr = np.arange(4, dtype=np.int32)
+    cuda_buf = ctx.buffer_from_data(arr)
+
+    arr = pa.Array.from_buffers(pa.int32(), 4, [None, cuda_buf])
+    buf_on_gpu = arr.buffers()[1]
+
+    assert buf_on_gpu.size == cuda_buf.size
+    assert buf_on_gpu.address == cuda_buf.address
+    assert buf_on_gpu.is_cpu == cuda_buf.is_cpu
+
+    repr1 = "<pyarrow.Buffer address="
+    repr2 = "size=16 is_cpu=False is_mutable=True>"
+    assert repr1 in repr(buf_on_gpu)
+    assert repr2 in repr(buf_on_gpu)
+
+    msg = "Implemented only for data on CPU device"
+    with pytest.raises(NotImplementedError, match=msg):
+        buf_on_gpu.hex()
+
+    with pytest.raises(NotImplementedError, match=msg):
+        cuda_buf.hex()
+
+    assert buf_on_gpu.is_mutable
+
+    with pytest.raises(NotImplementedError, match=msg):
+        buf_on_gpu[1]
+
+    with pytest.raises(NotImplementedError, match=msg):
+        cuda_buf[1]
+
+    with pytest.raises(NotImplementedError, match=msg):
+        pickle_module.dumps(buf_on_gpu, protocol=4)
+
+    buf = pa.py_buffer(b'testing')
+    arr = pa.FixedSizeBinaryArray.from_buffers(pa.binary(7), 1, [None, buf])
+    buf_on_gpu = arr.buffers()[1]
+    buf_on_gpu_sliced = buf_on_gpu.slice(2)
+
+    buf = pa.py_buffer(b'sting')
+    arr = pa.FixedSizeBinaryArray.from_buffers(pa.binary(5), 1, [None, buf])
+    buf_on_gpu_expected = arr.buffers()[1]
+
+    assert buf_on_gpu_sliced.equals(buf_on_gpu_expected)
+    assert buf.equals(buf_on_gpu_expected)
```

Comment:
I find it surprising that this passes for checking equality with CPU and non-CPU buffer ..

But can you also check `buf_on_gpu.equals(cuda_buf)` above? (so combining a Buffer and CudaBuffer object)

Reply:
Checking equation between `CudaBuffer` and `Buffer` on gpu work. But I get a segfault if checking equation between two `Buffers` on gpu (`buf_on_gpu_sliced` and `buf_on_gpu_expected`): https://github.com/apache/arrow/pull/41889/commits/958af940b3020f5656e86a08ee9f5d47d9857660

Reply:
`equals` doesn't seem to work. Example:

```python
>>> arr = np.array([b'testing'])
>>> cuda_buf = ctx.buffer_from_data(arr)
>>> arr = pa.FixedSizeBinaryArray.from_buffers(pa.binary(7), 1, [None, cuda_buf])
>>> buf_on_gpu = arr.buffers()[1]

>>> arr2 = np.array([b'testing'])
>>> cuda_buf2 = ctx.buffer_from_data(arr2)
>>> arr2 = pa.FixedSizeBinaryArray.from_buffers(pa.binary(7), 1, [None, cuda_buf2])
>>> buf_on_gpu2 = arr2.buffers()[1]

>>> buf_on_gpu.equals(buf_on_gpu2)
Segmentation fault
```

I am not sure if I can check that the `slice` method works correctly as I can't use the `to_pybytes()` or `equals()` methods. Comparing to cuda buffer also segfauls in this case:

```python
>>> arr = np.array([b'testing'])
>>> cuda_buf = ctx.buffer_from_data(arr)
>>> arr = pa.FixedSizeBinaryArray.from_buffers(pa.binary(7), 1, [None, cuda_buf])
>>> buf_on_gpu = arr.buffers()[1]

>>> buf_on_gpu.equals(cuda_buf)
True
>>> buf_on_gpu_sliced = buf_on_gpu.slice(2)

>>> arr_expected = np.array([b'sting'])
>>> cuda_buf_expected = ctx.buffer_from_data(arr_expected)

>>> buf_on_gpu_sliced.equals(cuda_buf_expected)
Segmentation fault
```

Reply:
> I am not sure if I can check that the `slice` method works correctly as I can't use the `to_pybytes()`

But to_pybytes should work on a CudaBuffer, I think?

> `equals` doesn't seem to work.

I also don't see this explicitly implemented for CudaBuffer on the C++ side (so if that uses the base Buffer::Equals implementation, this is probably expected to crash because it is comparing memory as if it is on the CPU). So let's protect from calling that for now on non-CPU, but should maybe open an issue about that whether it is worth implementing CudaBuffer::Equals.

Reply:
> But to_pybytes should work on a CudaBuffer, I think?

Yes, `to_pybytes` works correctly for CUDA buffer. But we are looking at the arrow buffer constructed from a CUDA one, right? So I can not check slicing for arrow buffer on gpu.

Reply:
Ah yes, but so you can always convert a Buffer with Cuda device to a CudaBuffer with `CudaBuffer.from_buffer(bug_on_gpu)`, and then you can use that method. 

Now, CudaBuffer.to_pybytes also just copies to the data to the host, so we should probably add similar logic for `Buffer.to_pybytes` (but this first requires to expose the copy methods to different device, so that's for a follow-up)

Reply:
Aha, didn't think of going back to CUDA buffer. Will do 👍 

> Now, CudaBuffer.to_pybytes also just copies to the data to the host, so we should probably add similar logic for `Buffer.to_pybytes` (but this first requires to expose the copy methods to different device, so that's for a follow-up)

Yes, that would be great!

Reply:
Issue opened: https://github.com/apache/arrow/issues/42053